### PR TITLE
Enable breakpoints and TS stack traces during test debugging

### PR DESCRIPTION
**Pull request by ckant**
## Description

- Upgrade mocha, vscode-test, and source-map-support

- Activate plugin before running unit tests

- Disable mocha test timeout
-  This allows for debugging without causing test timeouts

- Turn off coverage during test runs inside vscode

- Convert test runner to be closer to one on VSCode doc site
  - https://code.visualstudio.com/api/working-with-extensions/testing-extension

- Copy extension startup script to dist for testing

## Motivation and Context

- We were unable to debug the plugin from inside of tests (breakpoints weren't hit / was showing JS files / line numbers)
- Lazy loading of plugin during tests would cause random timeouts
- Debugging tests would cause the tests to timeout and fail
- Stack traces in test errors were showing JS files/line numbers

## Testing

Tests pass

## Checklist

- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] My code follows the code style of this project
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the changelog using the script `npm run newChange`

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/test/globalSetup.test.ts [hunk 1]
```diff
@@ -44,6 +45,9 @@ before(async () => {
     }
     const service = new DefaultTelemetryService(mockContext, mockAws, mockPublisher)
     ext.telemetry = service
+
+    // Warm up vscode before tests run
+    await vscode.extensions.getExtension(VSCODE_EXTENSION_ID.awstoolkit)!.activate()
```
Comment (justinmk3) [inline src/test/globalSetup.test.ts:50]: does this avoid a race?

Comment (ckant) [reply src/test/globalSetup.test.ts:50]: Yep, the first test that uses the extension activates it. That test often times out since activation takes a while.

**Comment by justinmk3** [comment]
codecov results suggest that maybe coverage is not running in CI?

**Comment by ckant** [comment]
> codecov results suggest that maybe coverage is not running in CI?

it seems to be reporting coverage for the test files themselves.

File: src/test/shared/defaultAwsClientBuilder.test.ts [hunk 2]
```diff
@@ -24,7 +24,7 @@ describe('DefaultAwsClientBuilder', () => {
 
             assert.strictEqual(!!service.config.customUserAgent, true)
             assert.strictEqual(
-                service.config.customUserAgent!,
+                service.config.customUserAgent!.replace('---Insiders', ''),
```
Comment (awschristou) [inline src/test/shared/defaultAwsClientBuilder.test.ts:27]: Why is this necessary?

File: src/test/globalSetup.test.ts (hunk 1 above)
Comment (awschristou) [reply src/test/globalSetup.test.ts:50]: Was this activation moved from somewhere else? This is interesting because not all tests require the extension to be activated

File: .vscode/launch.json [hunk 3]
```diff
@@ -41,14 +41,16 @@
...
                 "--extensionDevelopmentPath=${workspaceFolder}",
-                "--extensionTestsPath=${workspaceFolder}/dist/src/test/index.js"
+                "--extensionTestsPath=${workspaceFolder}/dist/src/test/index"
             ],
             "env": {
                 "TEST_REPORT_DIR": "${workspaceFolder}/.test-reports",
-                "TEST_FILE": null
+                "TEST_FILE": null,
+                "AWS_TOOLKIT_IGNORE_WEBPACK_BUNDLE": "true",
+                "NO_COVERAGE": "true"
```
[the remark this answers is not in the document]
Comment (ckant) [reply .vscode/launch.json:50]: Yeah, it used to hang the debug session even after the tests ran.

File: .vscode/launch.json [hunk 4]
```diff
@@ -41,14 +41,16 @@
...
             },
             "outFiles": ["${workspaceFolder}/dist/src/test/**/*.js"],
-            "preLaunchTask": "npm: watch"
+            "preLaunchTask": "${defaultBuildTask}"
```
[the remark this answers is not in the document]
Comment (ckant) [reply .vscode/launch.json:53]: There's an `isDefault` flag in the tasks.json file

```json
"group": {
  "kind": "build",
  "isDefault": true
}
```

File: src/test/testRunner.ts [hunk 5]
```diff
@@ -1,32 +1,251 @@
...
 
-    return new Promise<void>((resolve, reject) => {
-        legacyTestRunner.runTests(parameters.rootTestsPath, (err: any, failureCount: number) => {
+        const globalSetupPath = path.join(testsRoot, 'test', 'globalSetup.test.js')
+        if (testFilePath && fs.existsSync(globalSetupPath)) {
+            // XXX: explicitly add globalSetup, other tests depend on it.
+            mocha.addFile(globalSetupPath)
+        }
+
+        const testFolder = testType === 'unit' ? 'test' : 'integrationTest'
```
[the remark this answers is not in the document]
Comment (ckant) [reply src/test/testRunner.ts:57]: No real benefit. I just had it inside here and added this later to support integration tests.

File: src/test/shared/defaultAwsClientBuilder.test.ts (hunk 2 above)
Comment (ckant) [reply src/test/shared/defaultAwsClientBuilder.test.ts:27]: `ParentProduct` posts `Visual Studio Code - Insiders` instead of `Visual Studio Code` if you're using insiders.

File: src/test/testRunner.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (awschristou) [reply src/test/testRunner.ts:57]: I wonder if we'll wish we could pass paths in directly if/when we add more top level test groups/categories. 'testType ' doesn't appear to have any other value, so why bother?

File: .vscode/launch.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (awschristou) [reply .vscode/launch.json:53]: Hmm, is tasks.json a local development file? Doesn't seem like something we'd commit to the repo, which makes me wonder if we should revert that change.

File: src/test/shared/defaultAwsClientBuilder.test.ts (hunk 2 above)
Comment (awschristou) [reply src/test/shared/defaultAwsClientBuilder.test.ts:27]: Our other toolkits recognize the IDE variants being used. Does it cause an issue?

Comment (ckant) [reply src/test/shared/defaultAwsClientBuilder.test.ts:27]: It looks like we also report it in the `Parent Product Version`.

Product: `Visual Studio Code - Insiders`, Version: `1.45.0-insider`

Considering insiders is just a different build, I think we should probably combine them by stripping `- Insiders` from the user agent we post.

File: src/test/globalSetup.test.ts (hunk 1 above)
Comment (ckant) [reply src/test/globalSetup.test.ts:50]: I'd been consistently getting a single test timeout when running from the IDE. I believe that was because that specific test was the first to activate the extension and we had a timeout of a few seconds.

File: .vscode/launch.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (ckant) [reply .vscode/launch.json:53]: It's a part of the repo. It's what's responsible for running `npm watch / tsc watch` in the background automatically as soon as you open the project.

[the remark this answers is not in the document]
Comment (ckant) [reply .vscode/launch.json:53]: You can see it running if you run `Tasks: Show Running Tasks` from the command palette

File: src/test/testRunner.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (ckant) [reply src/test/testRunner.ts:57]: Sure, changed it to the test folder itself `test` or `integrationTest`.

File: src/test/testRunner.ts [hunk 6]
```diff
@@ -1,32 +1,256 @@
...
+    relativeCoverageDir: string
+    relativeSourcePath: string
+    ignorePatterns: string[]
+    includePid?: boolean
+    reports?: string[]
+    verbose?: boolean
+}
+
+function _readCoverOptions(testsRoot: string): TestRunnerOptions {
+    let coverConfigPath = path.join(testsRoot, '..', '..', 'coverconfig.json')
```
[the remark this answers is not in the document]
Comment (ckant) [reply src/test/testRunner.ts:248]: We were previously running from the `dist/src` directory. With this change, we are now running from the `dist` directory.

This was the root for the VSCode sample. I left it because it makes more sense since it emulates the packaged path.

packaged : `<root>/package.json`, `<root>/extensionMain.js`, `<root>/dist/extension.js`

test: `<root>/dist/package.json`, `<root>/dist/extensionMain.js`, `<root>/dist/src/extension.js`

The difference being that stuff in src gets webpacked and put directly into `dist`.

I'm pretty certain this could be made better, but this appears to work properly

**Comment by ckant** [comment]
Merging since I'd like to build off of this. We can discuss further iterations.

**Comment by justinmk3** [comment]
travis failure on master: https://travis-ci.org/github/aws/aws-toolkit-vscode/builds/682180650

```
rejected promise not handled within 1 second: TypeError: Cannot set property 'telemetryEnabled' of undefined
stack trace: TypeError: Cannot set property 'telemetryEnabled' of undefined
    at applyTelemetryEnabledState (/Users/travis/build/aws/aws-toolkit-vscode/dist/src/shared/telemetry/activation.js:9:5779)
    at Object.<anonymous> (/Users/travis/build/aws/aws-toolkit-vscode/dist/src/shared/telemetry/activation.js:9:4573)
    at Generator.next (<anonymous>)
    at /Users/travis/build/aws/aws-toolkit-vscode/dist/src/shared/telemetry/activation.js:9:1746
    at new Promise (<anonymous>)
    at __awaiter (/Users/travis/build/aws/aws-toolkit-vscode/dist/src/shared/telemetry/activation.js:9:662)
    at /Users/travis/build/aws/aws-toolkit-vscode/dist/src/shared/telemetry/activation.js:9:4183
    at e.fire (/Users/travis/build/aws/aws-toolkit-vscode/.vscode-test/vscode-1.44.2/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:48:497)
    at E.$acceptConfigurationChanged (/Users/travis/build/aws/aws-toolkit-vscode/.vscode-test/vscode-1.44.2/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:778:32)
    at /Users/travis/build/aws/aws-toolkit-vscode/.vscode-test/vscode-1.44.2/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:777:262
      ✓ detects running processes and successfully sends a kill signal to a running process - Unix (127ms)
      ✓ can not kill previously killed processes - Unix (226ms)
  CollectionUtils
```

Similar failures seen on CodeBuild in https://github.com/aws/aws-toolkit-vscode/pull/1070
